libs/core/src/lib/content-density/providers/content-density-observer.ts

```diff
@@ -0,0 +1,83 @@
+import { ChangeDetectorRef, ElementRef, Optional, Provider } from '@angular/core';
+import { map, Observable } from 'rxjs';
+import { DestroyedService } from '@fundamental-ngx/core/utils';
+import { GlobalContentDensityService } from '../services/global-content-density.service';
+import { ContentDensityObserverSettings, ContentDensityMode, LocalContentDensityMode } from '../content-density.types';
+import { CONTENT_DENSITY_DIRECTIVE } from '../tokens/content-density-directive';
+import { ContentDensityObserver } from '../classes/content-density-consumer.class';
+import { defaultContentDensityObserverConfigs } from '../variables/default-content-density-consumer-config';
+import { getChangesSource$ } from '../helpers/get-changes-source.provider';
+
+/**
+ * Creates provider for ContentDensityObserver
+ * @param providedConfiguration
+ */
+export function contentDensityObserver(providedConfiguration?: ContentDensityObserverSettings): Provider {
```

Comment:
we could have only one factory for all the consumers, something like this

```ts
export function consumerFactory(Consumer): Provider {
  return {
    provide: Consumer,
    useFactory: (injector: Injector) => new Consumer(injector),
    deps: [Injector],
  };
}
```

then, in our certain consumer we will use `injector.get(...)` to get any needed  stuff, so we would have something like that 

```ts
export class Consumer {
  constructor(private readonly injector: Injector) {
    // get custom token
    console.log(injector.get(Token));
  }

  consume(fn?: (injector: Injector) => any): void {
    if (fn) {
      // custom consumer logic
      fn(this.injector);
    } else {
      // default consumer logic
      this.injector
        .get(ElementRef)
        .nativeElement.classList.add('default-class');
    }
  }
}
```

very basic examples but i think logic is clear, let me know if not.

this way all the logic related to the consumer will be in the consumer itself and we won't need to create many factories. for sure some common logic for all the consumers may be moved to the factory (like destroyed service).

Reply:
Could be, but the way it is done right now seemed more explicit to me